lib/style/defs.ex

```diff
@@ -41,60 +41,107 @@ defmodule Styler.Style.Defs do
+      if def_meta[:do] do
+        # This is a def with a do block, like
+        #
+        #  def example(foo, bar \\ nil) do
+        #    :ok
+        #  end
+        #
+        def_line = def_meta[:line]
+        do_line = def_meta[:do][:line]
+        end_line = def_meta[:end][:line]
+        {:block, def_line, do_line, end_line}
+      else
+        # This is a def with a keyword do, like
+        #
+        #  def example(foo, bar \\ nil), do: :ok
+        #
+        [{{:__block__, do_meta, [:do]}, {_, body_meta, _}}] = body
+        def_line = def_meta[:line]
+        do_line = do_meta[:line]
+        end_line = body_meta[:closing][:line] || do_meta[:line]
+        {:keyword, def_line, do_line, end_line}
+      end
 
-    if def_start_line == def_end_line do
+    if def_line == end_line do
+      # Already collapsed
       {:skip, zipper, ctx}
     else
-      head = flatten_head(head, def_start_line)
-
-      {def_meta, body_meta_rewriter} =
-        if def_meta[:do] do
-          # we're in a `def do ... end`
-          delta = def_end_line - def_start_line
-          up_by_delta = &(&1 - delta)
+      delta = def_line - do_line
+      move_up = &(&1 + delta)
+      set_to_def_line = fn _ -> def_line end
 
-          # this is what does the shrinking of the `def ... do` stanza
+      {node, comments} =
+        if style == :block do
+          # We're working on a def do ... end
           def_meta =
             def_meta
-            |> Keyword.replace_lazy(:do, &Keyword.put(&1, :line, def_start_line))
-            |> Keyword.replace_lazy(:end, &Keyword.update!(&1, :line, up_by_delta))
+            |> Keyword.replace_lazy(:do, &Keyword.update!(&1, :line, set_to_def_line))
+            |> Keyword.replace_lazy(:end, &Keyword.update!(&1, :line, move_up))
 
-          # move all body line #s up by the amount we squished the head by
-          {def_meta, collapse_lines(up_by_delta)}
+          head = flatten_head(head, def_line)
+          body = update_all_meta(body, shift_lines(move_up))
+
+          comments =
+            ctx.comments
+            |> Style.displace_comments(def_line..do_line)
+            |> Style.shift_comments(do_line..end_line, delta)
+
+          {{def, def_meta, [head, body]}, comments}
         else
-          # we're in a `def, do:`
-          to_same_line = fn _ -> def_start_line end
-          {def_meta, collapse_lines(to_same_line)}
-        end
+          # We're working on a Keyword def do:
+          head = flatten_head(head, def_line)
+          body = update_all_meta(body, collapse_lines(set_to_def_line))
+          comments = Style.displace_comments(ctx.comments, def_line..end_line)
 
-      body = update_all_meta(body, body_meta_rewriter)
+          {{def, def_meta, [head, body]}, comments}
+        end
 
-      # There won't be any defs deeper in here, so lets skip ahead if we can
       # @TODO this skips checking the body, which can be incorrect if therey's a `quote do def do ...` inside of it
```

Comment:
lol this would be easily fixed if we just changed it from `skip` to `next`. 
¯\\_(ツ)_/¯
be smart to add tests for it tho. i can do it in a followup

Reply:
yeah i don't really have a lot of worries about macros that write code we should style not getting styled, because as we all know: the first rule of macros is, don't write macros